scheduler/scheduler.go

```diff
@@ -12,29 +12,33 @@ type Group struct {
 	interval *cronexpr.Expression
 }
 
-func (group Group) Schedule(ctx context.Context, runner func(), logger *util.Logger, logName string) {
+func (group Group) Schedule(ctx context.Context, runner func(context.Context), logger *util.Logger, logName string) {
 	go func() {
 		for {
-			delay := group.interval.Next(time.Now()).Sub(time.Now())
+			nextExecutions := group.interval.NextN(time.Now(), 2)
+			delay := time.Until(nextExecutions[0])
 
 			logger.PrintVerbose("Scheduled next run for %s in %+v", logName, delay)
 
 			select {
 			case <-ctx.Done():
 				return
 			case <-time.After(delay):
-				// NOTE: In the future we'll measure the runner's execution time
-				// and decide the next scheduling interval based on that
-				runner()
+				func() {
+					// Cancel runner at latest right before next scheduled execution should
+					// occur, to prevent skipping over runner executions by accident.
+					ctx, cancel := context.WithDeadline(ctx, nextExecutions[1].Add(-1*time.Second))
```

Comment:
Should we make this more conservative? If we're stuck in a slow query, our timeout is considerably larger than that...

Reply:
The hard part here is that I wanted to find something that also works for activity snapshots (every 10 seconds) - which wouldn't be the case if we made this e.g. 5 seconds.

Interestingly enough, the Go database driver will actually send a cancellation request to Postgres when the context timeout is reached (it seems from my testing), so we are not relying on the statement timeout in cases where the context has reached its deadline.

Reply:
Okay. We should check if this is still the case with our new driver if we migrate off `pq` at some point.